Add doctests and enable them

Erotemic opened the pull request:
I'm a big fan of doctests. Not only do they provide great documentation examples, they give developers a quick and easy way to get an entrypoint into a function, and you can run them to enhance your test suite and ensure examples continue to work.

In this PR I noticed something that was almost a doctest, and I turned it into a fully fledged one. It adds a little bit of boilerplate, but I think the ability to copy / paste the snippet into IPython and play with it is invaluable.

I also added a doctest to `parse_object_spec`, that I wrote when I was playing with it to understand it. 

In addition to adding these doctests here I added `--doctest-modules` to the default pytest config in `setup.cfg` (which should probably be ported to fully use `pyproject.toml`, but that will be a different PR). This will enable these tests on the CI.

As I continue to use / learn about this repo I'm likely to add more doctests as I debug / understand. This PR ensures that they will be tested if I push them upstream.

For now I'm using the stdlib doctest, but in the future we could use [xdoctest](https://github.com/Erotemic/xdoctest) if we want a more flexible doctest syntax.

yifanmai commented:
I agree in principle with using doctests, but this pull requests causes a large number of test failures due to existing comments being interpreted as doctests.

Erotemic commented:
A lot of these can be fixed with xdoctest. Let me see what I can do about it.

Erotemic commented:
@yifanmai I was able to fix the errors by switching to xdoctest. Unlike the stdlib doctest module it uses static parsing to extract docstrings, which means modules (which might have unmet dependencies) that don't have any enabled doctests will never be imported, and thus won't cause import errors. Additionally I've set `--xdoctest-style=google` in the pytest arguments, which means it will only parse doctests indented under an `Example:` block in the docstring.

Additionally I've fixed string escape errors in `src/helm/benchmark/scenarios/math_scenario.py`, and the other non-doctest related string escape errors are fixed in a separate PR: https://github.com/stanford-crfm/helm/pull/3751

The other change is that I noticed `install-dev.sh` was not installing the test dependencies like pytest (and now xdoctest). Instead it was relying on the pinned requirements.txt file to do that. So I fixed that to install both `[all,dev]` instead of just `[all]`, so now if a new test dependency is added, the CI requirements.txt updater should include it without needing to be explicitly included (which I also did here).

yifanmai commented:
Thanks! Yes, the dev extras should be installed in `install-dev.sh`. The additional changes look good to me.
